src/sds.c

```diff
@@ -65,6 +55,32 @@ static inline size_t sdsTypeMaxSize(char type) {
     return -1; /* this is equivalent to the max SDS_TYPE_64 or SDS_TYPE_32 */
 }
 
+/* 
+ * Adjusts the SDS type if the allocated buffer size exceeds the maximum size 
+ * addressable by the current type.
+ *
+ * The SDS type is initially determined based on the logical length of the string. 
+ * However, allocators like jemalloc may return a buffer larger than requested, 
+ * potentially exceeding the maximum size the selected SDS type can handle. This 
+ * can cause a mismatch between the `alloc` field and the actual buffer size, 
+ * leading to wasted memory and possible inconsistencies.
+ *
+ * This function ensures that the SDS type is selected based on the actual buffer 
+ * size rather than just the logical length. If the buffer size supports a larger 
+ * SDS type, it updates `type` and `hdrlen` accordingly.
+ *
+ * Returns 1 if the type was adjusted, 0 otherwise.
+ */
+static inline int adjustTypeIfNeeded(char *type, int *hdrlen, size_t bufsize) {
+    size_t usable = bufsize - *hdrlen - 1;
+    if (*type != SDS_TYPE_5 && usable > sdsTypeMaxSize(*type)) {
```

Comment:
why can't the type of SDS_TYPE_5 to be changed?

Reply:
and i found that this block are not covered by test in https://app.codecov.io/github/redis/redis/blob/moticless%2Fredis%3Aunify-keys-and-vals/src%2Fsds.c